opentracing/recorder.go

```diff
@@ -0,0 +1,90 @@
+package opentracing
+
+import (
+	"fmt"
+	"log"
+	"sync"
+
+	basictracer "github.com/opentracing/basictracer-go"
+	"sourcegraph.com/sourcegraph/appdash"
+)
+
+// Recorder implements basictracer.Recorder.
+type Recorder struct {
+	collector appdash.Collector
+	logOnce   sync.Once
+	verbose   bool
+}
+
+// NewRecorder forwards basictracer.RawSpans to an appdash.Collector.
+func NewRecorder(collector appdash.Collector, opts Options) *Recorder {
+	return &Recorder{
+		collector: collector,
+		verbose:   opts.Verbose,
+	}
+}
+
+// RecordSpan converts a RawSpan into the Appdash representation of a span
+// and records it to the underlying collector.
+func (r *Recorder) RecordSpan(sp basictracer.RawSpan) {
+	spanID := appdash.SpanID{
+		Span:   appdash.ID(uint64(sp.SpanID)),
+		Trace:  appdash.ID(uint64(sp.TraceID)),
+		Parent: appdash.ID(uint64(sp.ParentSpanID)),
+	}
+
+	r.collectEvent(spanID, appdash.SpanName(sp.Operation))
+
+	// Record all of the logs. Payloads are thrown out.
+	for _, log := range sp.Logs {
+		r.collectEvent(spanID, appdash.LogWithTimestamp(log.Event, log.Timestamp))
+	}
+
+	for key, value := range sp.Tags {
+		val := []byte(fmt.Sprintf("%v", value))
+		r.collectAnnotation(spanID, appdash.Annotation{Key: key, Value: val})
+	}
+
+	for key, val := range sp.Baggage {
+		r.collectAnnotation(spanID, appdash.Annotation{Key: key, Value: []byte(val)})
+	}
+
+	// Add the duration to the start time to get an approximate end time.
+	approxEndTime := sp.Start.Add(sp.Duration)
+	r.collectEvent(spanID, appdash.Timespan{S: sp.Start, E: approxEndTime})
+}
+
+// collectEvent marshals and collects the Event.
+func (r *Recorder) collectEvent(spanID appdash.SpanID, e appdash.Event) {
+	ans, err := appdash.MarshalEvent(e)
+	if err != nil {
+		r.logError(spanID, err)
+		return
+	}
+	r.collectAnnotation(spanID, ans...)
+}
+
+func (r *Recorder) collectAnnotation(spanID appdash.SpanID, ans ...appdash.Annotation) {
+	err := r.collector.Collect(spanID, ans...)
+	if err != nil {
+		r.logError(spanID, err)
+	}
+}
+
+// logError converts an error into a log event and collects it.
+// If for whatever reason the error can't be collected, it is logged to stdout.
+func (r *Recorder) logError(spanID appdash.SpanID, err error) {
+	ans, _ := appdash.MarshalEvent(appdash.Log(err.Error()))
+	collectErr := r.collector.Collect(spanID, ans...)
+
+	// At this point, something is definitely wrong.
+	if collectErr != nil {
+		if r.verbose {
+			log.Printf("opentracing appdash error: %v\n", collectErr)
+		} else {
+			r.logOnce.Do(func() {
+				log.Printf("OpenTracing Appdash error: %v\n", collectErr)
```

Comment:
Better to let the user specify the logger, if desired. E.g. see how we do this in https://github.com/sourcegraph/appdash/blob/master/collector.go#L105-L107

The default logger would then be set in NewRecorder.